zebra-chain/src/commitments/sapling.rs

```diff
@@ -0,0 +1,291 @@
+    fn from(extended_point: jubjub::ExtendedPoint) -> Self {
+        Self(jubjub::AffinePoint::from(extended_point))
+    }
+}
+
+impl From<NoteCommitment> for [u8; 32] {
+    fn from(cm: NoteCommitment) -> [u8; 32] {
+        cm.0.to_bytes()
+    }
+}
+
+impl Eq for NoteCommitment {}
+
+impl ZcashSerialize for NoteCommitment {
+    fn zcash_serialize<W: io::Write>(&self, mut writer: W) -> Result<(), io::Error> {
+        writer.write_all(&self.0.to_bytes())?;
+        Ok(())
+    }
+}
+
+impl ZcashDeserialize for NoteCommitment {
+    fn zcash_deserialize<R: io::Read>(mut reader: R) -> Result<Self, SerializationError> {
+        Ok(Self(
+            jubjub::AffinePoint::from_bytes(reader.read_32_bytes()?).unwrap(),
+        ))
+    }
+}
+
+impl NoteCommitment {
+    /// Generate a new _NoteCommitment_ and the randomness used to create it.
+    ///
+    /// We return the randomness because it is needed to construct a _Note_,
+    /// before it is encrypted as part of an _Output Description_.
+    ///
+    /// NoteCommit^Sapling_rcm (g*_d , pk*_d , v) :=
+    ///   WindowedPedersenCommit_rcm([1; 6] || I2LEBSP_64(v) || g*_d || pk*_d)
+    ///
+    /// https://zips.z.cash/protocol/protocol.pdf#concretewindowedcommit
+    #[allow(non_snake_case)]
+    pub fn new<T>(
+        csprng: &mut T,
+        diversifier: Diversifier,
+        transmission_key: TransmissionKey,
+        value: Amount<NonNegative>,
+    ) -> (CommitmentRandomness, Self)
+    where
+        T: RngCore + CryptoRng,
+    {
+        // s as in the argument name for WindowedPedersenCommit_r(s)
+        let mut s: BitVec<Lsb0, u8> = BitVec::new();
+
+        // Prefix
+        s.append(&mut bitvec![1; 6]);
+
+        // Jubjub repr_J canonical byte encoding
+        // https://zips.z.cash/protocol/protocol.pdf#jubjub
+        let g_d_bytes = jubjub::AffinePoint::from(diversifier).to_bytes();
+        let pk_d_bytes = <[u8; 32]>::from(transmission_key);
+        let v_bytes = value.to_bytes();
+
+        s.append(&mut BitVec::<Lsb0, u8>::from_slice(&g_d_bytes[..]));
+        s.append(&mut BitVec::<Lsb0, u8>::from_slice(&pk_d_bytes[..]));
+        s.append(&mut BitVec::<Lsb0, u8>::from_slice(&v_bytes[..]));
+
+        let rcm = CommitmentRandomness(generate_trapdoor(csprng));
+
+        (
+            rcm,
+            NoteCommitment::from(windowed_pedersen_commitment(rcm.0, &s)),
+        )
+    }
+
+    /// Hash Extractor for Jubjub (?)
+    ///
+    /// https://zips.z.cash/protocol/protocol.pdf#concreteextractorjubjub
+    pub fn extract_u(&self) -> jubjub::Fq {
+        self.0.get_u()
+    }
+}
+
+/// A Homomorphic Pedersen commitment to the value of a note, used in Spend and
+/// Output Descriptions.
+///
+/// https://zips.z.cash/protocol/protocol.pdf#concretehomomorphiccommit
+#[derive(Clone, Deserialize, PartialEq, Serialize)]
+pub struct ValueCommitment(#[serde(with = "serde_helpers::AffinePoint")] pub jubjub::AffinePoint);
+
+impl fmt::Debug for ValueCommitment {
```

Comment:
Just checking: the reason for a handcrafted Debug impl is because the derived impl calling the upstream Debug impl on AffinePoint doesn't work well, or?

(not an objection, just wondering if there's something that should ideally be fixed upstream since it's easy to lose track)

Reply:
I think this is to keep it consistent with other hex-based Debug's, else it's an array of byte values instead of a nice hex output